pkg/aks/create.go

```diff
@@ -23,17 +24,159 @@ func CreateResourceGroup(ctx context.Context, groupsClient *resources.GroupsClie
 	return err
 }
 
-// CreateOrUpdateCluster creates a new managed Kubernetes cluster
-func CreateOrUpdateCluster(ctx context.Context, cred *Credentials, clusterClient *containerservice.ManagedClustersClient,
+// CreateCluster creates a new managed Kubernetes cluster. In this case, there will be no existing upstream cluster.
+// We are provisioning a brand new one.
+func CreateCluster(ctx context.Context, cred *Credentials, clusterClient *containerservice.ManagedClustersClient,
 	spec *aksv1.AKSClusterConfigSpec, phase string) error {
 
+	managedCluster, err := CreateManagedCluster(ctx, cred, spec, phase)
+	if err != nil {
+		return err
+	}
+
+	_, err = clusterClient.CreateOrUpdate(
+		ctx,
+		spec.ResourceGroup,
+		spec.ClusterName,
+		managedCluster,
+	)
+
+	return err
+}
+
+// UpdateCluster updates an existing managed Kubernetes cluster. Before updating, it pulls any existing configuration
+// and then only updates managed fields.
+func UpdateCluster(ctx context.Context, cred *Credentials, clusterClient *containerservice.ManagedClustersClient,
+	spec *aksv1.AKSClusterConfigSpec, phase string) error {
+
+	// Create a new managed cluster from the AKS cluster config
+	managedCluster, err := CreateManagedCluster(ctx, cred, spec, phase)
+	if err != nil {
+		return err
+	}
+
+	// Pull the upstream cluster state
+	aksCluster, err := clusterClient.Get(ctx, spec.ResourceGroup, spec.ClusterName)
+	if err != nil {
+		logrus.Errorf("Error getting upstream AKS cluster by name [%s]: %s", spec.ClusterName, err.Error())
+		return err
+	}
+
+	// Upstream cluster state was successfully pulled. Merge in updates without overwriting upstream fields: we never
+	// want to overwrite preconfigured values in Azure with nil values. So only update fields pulled from AKS with
+	// values from the managed cluster if they are non nil.
+
+	/*
+	The following fields are managed in Rancher but are NOT configurable on update
+	- Name
+	- Location
+	- DNSPrefix
+	- EnablePrivateCluster
+	- LoadBalancerProfile
+	 */
+
+	// Update kubernetes version
+	aksCluster.KubernetesVersion = managedCluster.KubernetesVersion
```

Comment:
If a cluster is imported, we might not have kubernetes version set on the spec in Rancher. Is there a risk this could override the upstream value with a nil value?